plugins/dell-dock2/fu-dell-dock2-ec.c

```diff
@@ -0,0 +1,972 @@
+			return FALSE;
+	}
+
+	/* Intel i266-LM */
+	if (fu_dell_dock2_ec_dev_entry(device, DELL_DOCK2_EC_DEV_TYPE_LAN, 0, 0) != NULL) {
+		g_autoptr(FuDellDock2Ilan) ilan_device = NULL;
+
+		ilan_device = fu_dell_dock2_ilan_new(device);
+		if (!fu_dell_dock2_ec_create_node(device, FU_DEVICE(ilan_device), error))
+			return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_ec_dock_type_extract(FuDevice *device, GError **error)
+{
+	FuDellDock2BaseType dock_type = fu_dell_dock2_ec_get_dock_type(device);
+	guint8 dev_type = DELL_DOCK2_EC_DEV_TYPE_MAIN_EC;
+
+	/* don't change error type, the plugin ignores it */
+	if (dock_type != FU_DELL_DOCK2_BASE_TYPE_K2) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND, "No valid dock was found");
+		return FALSE;
+	}
+
+	/* this will trigger setting up all the quirks */
+	fu_device_add_instance_u8(device, "DOCKTYPE", dock_type);
+	fu_device_add_instance_u8(device, "DEVTYPE", dev_type);
+	fu_device_build_instance_id(device,
+				    error,
+				    "USB",
+				    "VID",
+				    "PID",
+				    "DOCKTYPE",
+				    "DEVTYPE",
+				    NULL);
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_ec_dock_type_cmd(FuDevice *device, GError **error)
+{
+	FuDellDock2Ec *self = FU_DELL_DOCK2_EC(device);
+	gsize length = 1;
+	g_autoptr(GByteArray) res = g_byte_array_new_take(g_malloc0(length), length);
+	guint8 cmd = DELL_DOCK2_EC_HID_CMD_GET_DOCK_TYPE;
+
+	/* expect response 1 byte */
+	if (!fu_dell_dock2_ec_read(device, cmd, res, error)) {
+		g_prefix_error(error, "Failed to query dock type: ");
+		return FALSE;
+	}
+
+	self->base_type = res->data[0];
+
+	/* check dock type to proceed with this plugin or exit as unsupported */
+	return fu_dell_dock2_ec_dock_type_extract(device, error);
+}
+
+static gboolean
+fu_dell_dock2_ec_dock_info_extract(FuDevice *device, GError **error)
+{
+	FuDellDock2Ec *self = FU_DELL_DOCK2_EC(device);
+
+	if (!self->dock_info) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "Failed to parse dock info");
+		return FALSE;
+	}
+
+	if (self->dock_info->header.total_devices == 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_SIGNATURE_INVALID,
+			    "No bridge devices detected, dock may be booting up");
+		return FALSE;
+	}
+	g_info("found %u devices [%u->%u]",
+	       self->dock_info->header.total_devices,
+	       self->dock_info->header.first_index,
+	       self->dock_info->header.last_index);
+
+	for (guint i = 0; i < self->dock_info->header.total_devices; i++) {
+		struct FuDellDock2EcQueryEntry dev_entry = self->dock_info->devices[i];
+		const gchar *type_str;
+		const gchar *location_str;
+		guint32 version32 = GUINT32_FROM_BE(dev_entry.version.version_32);
+
+		/* name the component */
+		type_str = fu_dell_dock2_ec_devicetype_to_str(dev_entry.ec_addr_map.device_type,
+							      dev_entry.ec_addr_map.sub_type,
+							      dev_entry.ec_addr_map.instance);
+		if (type_str == NULL)
+			continue;
```

Comment:
this seems like a pretty bad problem if you hit this flow.  Should it be a warning?

Reply:
Fixed, thanks.